test/test_fluid.py

```diff
@@ -0,0 +1,162 @@
+from meshmode.array_context import (  # noqa
+    pytest_generate_tests_for_pyopencl_array_context
+    as pytest_generate_tests)
+
+logger = logging.getLogger(__name__)
+
+
+@pytest.mark.parametrize("dim", [1, 2, 3])
+def test_velocity_gradient(actx_factory, dim):
+    """Test that the velocity gradient does the right things."""
+    from mirgecom.fluid import compute_local_velocity_gradient
+    actx = actx_factory()
+
+    nel_1d = 16
+
+    from meshmode.mesh.generation import generate_regular_rect_mesh
+
+    mesh = generate_regular_rect_mesh(
+        a=(1.0,) * dim, b=(2.0,) * dim, n=(nel_1d,) * dim
+    )
+
+    order = 3
+    discr = EagerDGDiscretization(actx, mesh, order=order)
+    nodes = thaw(actx, discr.nodes())
+
+    zeros = discr.zeros(actx)
+    ones = zeros + 1.0
+
+    logger.info(f"Number of {dim}d elems: {mesh.nelements}")
+
+    # Sanity check: grad(v=0) == 0
+    mass = ones
+    energy = zeros + 2.5
+    mom = make_obj_array([zeros for _ in range(dim)])
+    q = join_conserved(dim, mass=mass, energy=energy, momentum=mom)
+
+    cv = split_conserved(dim, q)
+    grad_v = compute_local_velocity_gradient(discr, cv)
+
+    grad_v_norm = [discr.norm(grad_v[i], np.inf) for i in range(dim)]
+    tol = 1e-16
+    for i in range(dim):
+        assert grad_v_norm[i] < tol
+
+    # Sanity check: grad_j(v_i=r_i) == I
+    mom = nodes
+    q = join_conserved(dim, mass=mass, energy=energy, momentum=mom)
+    cv = split_conserved(dim, q)
+    grad_v = compute_local_velocity_gradient(discr, cv)
+    tol = 1e-12
+    for i in range(dim):
+        grad_v_comp = grad_v[i]
+        for j in range(dim):
+            if i == j:
+                exp_value = 1.0
+            else:
+                exp_value = 0.0
+            assert discr.norm(grad_v_comp[j] - exp_value, np.inf) < tol
+
+    # Sanity check: grad_j(v_i=r_i) == I, constant rho != 1.0
+    mass = zeros + 2.0
+    mom = mass*nodes
+    q = join_conserved(dim, mass=mass, energy=energy, momentum=mom)
+    cv = split_conserved(dim, q)
+    grad_v = compute_local_velocity_gradient(discr, cv)
+    tol = 1e-12
+    for i in range(dim):
+        grad_v_comp = grad_v[i]
+        for j in range(dim):
+            if i == j:
+                exp_value = 1.0
+            else:
+                exp_value = 0.0
+            assert discr.norm(grad_v_comp[j] - exp_value, np.inf) < tol
+
+    # Sanity check: grad_j(v_i=r_i) == I, spatially varying rho
+    mass = ((nodes[0] + 2.0) * nodes[0])  # quadratic rho
+    mom = mass*nodes
+    q = join_conserved(dim, mass=mass, energy=energy, momentum=mom)
+    cv = split_conserved(dim, q)
+    grad_v = compute_local_velocity_gradient(discr, cv)
+    tol = 1e-12
+    for i in range(dim):
+        grad_v_comp = grad_v[i]
+        for j in range(dim):
+            if i == j:
+                exp_value = 1.0
+            else:
+                exp_value = 0.0
+            assert discr.norm(grad_v_comp[j] - exp_value, np.inf) < tol
```

Comment:
Possible to avoid test code duplication by parametrizing the test on `cv` and `expected_grad_v` (or functions that generate them given a discretization, etc.)?

Reply:
(f3d0c98)